--- sys/arch/amd64/pci/acpipci.c
+++ sys/arch/amd64/pci/acpipci.c
@@ -194,23 +194,29 @@
 	    (sc->sc_acpi->sc_fadt->iapc_boot_arch & FADT_NO_MSI) == 0)
 		pba.pba_flags |= PCI_FLAGS_MSI_ENABLED;
 
-	/* Enable MSI for QEMU claiming ACPI 1.0 */
+	/* Enable MSI for QEMU or Amazon EC2 claiming ACPI 1.0 */
 	tag = pci_make_tag(pba.pba_pc, sc->sc_bus, 0, 0);
 	id = pci_conf_read(pba.pba_pc, tag, PCI_SUBSYS_ID_REG);
 	if (sc->sc_acpi->sc_fadt->hdr.revision == 1 &&
-	    PCI_VENDOR(id) == PCI_VENDOR_QUMRANET)
+	    (PCI_VENDOR(id) == PCI_VENDOR_QUMRANET ||
+	     PCI_VENDOR(id) == PCI_VENDOR_AMAZON))
+		pba.pba_flags |= PCI_FLAGS_MSI_ENABLED;
+
+	id = pci_conf_read(pba.pba_pc, tag, PCI_ID_REG);
+	if (sc->sc_acpi->sc_fadt->hdr.revision == 1 &&
+	    PCI_VENDOR(id) == PCI_VENDOR_AMAZON)
 		pba.pba_flags |= PCI_FLAGS_MSI_ENABLED;
 
 	/*
 	 * Don't enable MSI on chipsets from low-end manufacturers
 	 * like VIA and SiS.  We do this by looking at the host
 	 * bridge, which should be device 0 function 0.
 	 */
-	id = pci_conf_read(pba.pba_pc, tag, PCI_ID_REG);
 	class = pci_conf_read(pba.pba_pc, tag, PCI_CLASS_REG);
 	if (PCI_CLASS(class) == PCI_CLASS_BRIDGE &&
 	    PCI_SUBCLASS(class) != PCI_SUBCLASS_BRIDGE_HOST &&
 	    PCI_VENDOR(id) != PCI_VENDOR_AMD &&
 	    PCI_VENDOR(id) != PCI_VENDOR_NVIDIA &&
-	    PCI_VENDOR(id) != PCI_VENDOR_INTEL)
+	    PCI_VENDOR(id) != PCI_VENDOR_INTEL &&
+	    PCI_VENDOR(id) != PCI_VENDOR_AMAZON)
 		pba.pba_flags &= ~PCI_FLAGS_MSI_ENABLED;
